scripts/gen_idt.py

```diff
@@ -0,0 +1,250 @@
+    sys.stdout.write(os.path.basename(sys.argv[0]) + ": " + text + "\n")
+
+def error(text):
+    sys.stderr.write(os.path.basename(sys.argv[0]) + ": " + text + "\n")
+    sys.exit(1)
+
+# See Section 6.11 of the Intel Architecture Software Developer's Manual
+gate_desc_format = "<HHBBH"
+
+def create_irq_gate(handler, dpl):
+    present = 1
+    gate_type = 0xE # 32-bit interrupt gate
+    type_attr = gate_type | (dpl << 5) | (present << 7)
+
+    offset_hi = handler >> 16
+    offset_lo = handler & 0xFFFF
+
+    data = struct.pack(gate_desc_format, offset_lo, KERNEL_CODE_SEG, 0,
+            type_attr, offset_hi)
+    return data
+
+
+def create_task_gate(tss, dpl):
+    present = 1
+    gate_type = 0x5 # 32-bit task gate
+    type_attr = gate_type | (dpl << 5) | (present << 7)
+
+    data = struct.pack(gate_desc_format, 0, tss, 0, type_attr, 0)
+    return data
+
+
+def create_idt_binary(idt_config, filename):
+    with open(filename, "wb") as fp:
+        for handler, tss, dpl in idt_config:
+            if handler and tss:
+                error("entry specifies both handler function and tss")
+
+            if not handler and not tss:
+                error("entry does not specify either handler or tss")
+
+            if (handler):
+                data = create_irq_gate(handler, dpl)
+            else:
+                data = create_task_gate(tss, dpl)
+
+            fp.write(data)
+
+map_fmt = "<B"
+
+def create_irq_vec_map_binary(irq_vec_map, filename):
+    with open(filename, "wb") as fp:
+        for i in irq_vec_map:
+            fp.write(struct.pack(map_fmt, i))
+
+def priority_range(prio):
+    # Priority levels are represented as groups of 16 vectors within the IDT
+    base = 32 + (prio * 16)
+    return range(base, base + 16)
+
+
+def update_irq_vec_map(irq_vec_map, irq, vector, max_irq):
+    # No IRQ associated; exception or software interrupt
+    if irq == -1:
+        return
+
+    if irq >= max_irq:
+        error("irq %d specified, but CONFIG_MAX_IRQ_LINES is %d" %
+                (irq, max_irq))
+
+    # This table will never have values less than 32 since those are for
+    # exceptions; 0 means unconfigured
+    if irq_vec_map[irq] != 0:
+        error("multiple vector assignments for interrupt line %d", irq)
+
+    debug("assign IRQ %d to vector %d" % (irq, vector))
+    irq_vec_map[irq] = vector
+
+
+def setup_idt(spur_code, spur_nocode, intlist, max_vec, max_irq):
+    irq_vec_map = [0 for i in range(max_irq)]
+    vectors = [None for i in range(max_vec)]
+
+    # Pass 1: sanity check and set up hard-coded interrupt vectors
+    for handler, irq, prio, vec, dpl, tss in intlist:
+        if vec == -1:
+            if prio == -1:
+                error("entry does not specify vector or priority level")
+            continue
+
+        if (vec >= max_vec):
+            error("Vector %d specified, but size of IDT is only %d vectors" %
+                    (vec, max_vec))
+
+        if (vectors[vec] != None):
```

Comment:
I guess this is what happens when you code switch between C and Python many times in a single day